backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt

```diff
@@ -78,55 +74,93 @@ internal val IrClass.isArrayWithFixedSizeItems: Boolean
-    }
+internal fun IrFunction.hasReferenceAt(index: Int) = when (index) {
+    BridgeDirection.RETURN_INDEX -> isSuspend || returnType.let { !it.isInlinedNative() && !it.isUnit() }
+    BridgeDirection.DISPATCH_RECEIVER_INDEX -> dispatchReceiverParameter.let { it != null && !it.type.isInlinedNative() }
+    BridgeDirection.EXTENSION_RECEIVER_INDEX -> extensionReceiverParameter.let { it != null && !it.type.isInlinedNative() }
+    else -> !this.valueParameters[BridgeDirection.unmapParameterIndex(index)].type.isInlinedNative()
+}
+
+private fun IrFunction.typeAt(index: Int) = when (index) {
+    BridgeDirection.RETURN_INDEX -> if (isSuspend) null else returnType
+    BridgeDirection.DISPATCH_RECEIVER_INDEX -> dispatchReceiverParameter?.type
+    BridgeDirection.EXTENSION_RECEIVER_INDEX -> extensionReceiverParameter?.type
+    else -> this.valueParameters[BridgeDirection.unmapParameterIndex(index)].type
 }
 
 private fun IrFunction.needBridgeToAt(target: IrFunction, index: Int)
-        = hasValueTypeAt(index) xor target.hasValueTypeAt(index)
+        = bridgeDirectionToAt(target, index).kind != BridgeDirectionKind.NONE
 
 internal fun IrFunction.needBridgeTo(target: IrFunction)
         = (0..this.valueParameters.size + 2).any { needBridgeToAt(target, it) }
 
-internal enum class BridgeDirection {
-    NOT_NEEDED,
-    FROM_VALUE_TYPE,
-    TO_VALUE_TYPE
+internal enum class BridgeDirectionKind {
+    NONE,
+    BOX,
+    UNBOX
+}
+
+internal data class BridgeDirection(val irClass: IrClass?, val kind: BridgeDirectionKind) {
+    companion object {
+        val NONE = BridgeDirection(null, BridgeDirectionKind.NONE)
+        const val RETURN_INDEX = 0
+        const val DISPATCH_RECEIVER_INDEX = 1
+        const val EXTENSION_RECEIVER_INDEX = 2
+        fun mapParameterIndex(index: Int) = index + 3
+        fun unmapParameterIndex(index: Int) = index - 3
+    }
+}
+
+internal fun IrType.isNullableNothing() =
+        isNullable() && classifierOrNull?.isClassWithFqName(KotlinBuiltIns.FQ_NAMES.nothing) == true
+
+private fun IrFunction.bridgeDirectionToAt(overriddenFunction: IrFunction, index: Int): BridgeDirection {
+    val irClass = overriddenFunction.typeAt(index)?.erasure()
+    return when {
+        index == 0 && returnType.isNothing() && !overriddenFunction.returnType.isNothing() ->
+            BridgeDirection(irClass, BridgeDirectionKind.UNBOX)
+
+        index == 0 && returnType.isNullableNothing()
+                && overriddenFunction.returnType.computePrimitiveBinaryTypeOrNull() == PrimitiveBinaryType.POINTER ->
+            BridgeDirection(irClass, BridgeDirectionKind.UNBOX)
+
+        hasValueTypeAt(index) && overriddenFunction.hasReferenceAt(index) ->
+            // Erase to [Any?].
+            BridgeDirection(null, BridgeDirectionKind.BOX)
+
+        hasReferenceAt(index) && overriddenFunction.hasValueTypeAt(index) ->
+            BridgeDirection(irClass, BridgeDirectionKind.UNBOX)
+
+        else -> BridgeDirection.NONE
+    }
```

Comment:
Why not reuse the logic from `getTypeConversion`?

Reply:
I don't see any handling of `Nothing` there.